lib/amazonka/src/Amazonka/Auth/SSO.hs

```diff
@@ -0,0 +1,165 @@
+import Amazonka.Auth.ConfigFile (configPathRelative, loadIniFile)
+import Amazonka.Auth.Exception
+import Amazonka.Data.Sensitive
+import Amazonka.Data.Time (ISO8601, Time (..))
+import Amazonka.Env (Env, Env' (..), lookupRegion)
+import Amazonka.HTTP (retryRequest)
+import Amazonka.Lens (throwingM, (^.))
+import Amazonka.Prelude
+import Amazonka.SSO.GetRoleCredentials as SSO
+import qualified Amazonka.SSO.Types as SSO (RoleCredentials (..))
+import Amazonka.Types
+import qualified Control.Exception as Exception
+import Control.Monad.Trans.Resource (runResourceT)
+import Data.Aeson (FromJSON, decodeFileStrict)
+import qualified Data.HashMap.Strict as HashMap
+import Data.Text (pack)
+import Data.Text.Encoding (encodeUtf8)
+import Data.Time.Clock.POSIX (posixSecondsToUTCTime)
+import qualified Network.HTTP.Client as Client
+import qualified System.Directory as Directory
+import qualified System.Environment as Environment
+import System.FilePath.Glob (decompile, globDir1)
+
+-- | Data present in @~/.aws/sso/cache/x.json@
+data Cache = Cache
+  { accessToken :: Text,
+    region :: Region
+  }
+  deriving stock (Generic)
+  deriving anyclass (FromJSON)
+
+-- | Data present in @~/.aws/config@
+data Config = Config
+  { sso_account_id :: Text,
+    sso_role_name :: Text
+  }
+
+fromSSOCache ::
+  ( MonadIO m,
+    Foldable withAuth
+  ) =>
+  Env' withAuth ->
+  m Env
+fromSSOCache env = do
+  auth <- liftIO $ fetchAuthInBackground getCredentials
+  discoverRegion <&> \case
+    Nothing -> env {envAuth = Identity auth}
+    Just r -> env {envRegion = r, envAuth = Identity auth}
+  where
+    getCredentials = do
+      Cache {..} <- readCache
+      Config {..} <- readConfig
+
+      -- Use the SSO region, even if configuring for another
+      let ssoEnv = env {envRegion = region}
+
+      resp <-
+        runResourceT $
+          sendWithoutAuth ssoEnv $
+            SSO.newGetRoleCredentials sso_role_name sso_account_id accessToken
+
+      let mCreds = do
+            rc <- resp ^. SSO.getRoleCredentialsResponse_roleCredentials
+            roleCredentialsToAuthEnv rc
+
+      case mCreds of
+        Nothing -> fail "sso:GetRoleWithCredentials returned no credentials."
+        Just c -> pure c
+
+-- | Send an unsigned request and read the response body
+sendWithoutAuth ::
+  ( MonadResource m,
+    AWSRequest a,
+    Foldable withAuth
+  ) =>
+  Env' withAuth ->
+  a ->
+  m (AWSResponse a)
+sendWithoutAuth env req = do
+  eResponse <- retryRequest env req
+  either (liftIO . Exception.throwIO) (pure . Client.responseBody) eResponse
+
+-- | Read @~/.aws/sso/cache/*.json@
+--
+-- This will throw @_MissingEnvError@ if there is no such file or none that
+-- correctly parse; otherwise it returns the first -- though we never expect
+-- there to be more than one (even if you've @sso login@d multiple profiles).
+readCache :: MonadIO m => m Cache
+readCache = liftIO $ do
+  home <- Directory.getHomeDirectory
+  paths <- globDir1 glob home
```

Comment:
From that botocore snippet I put on the linked issue, we might be able to guess the name of the config file if we can find the right URL to hash? (If we can't, then globbing is fine)